File: mobile-app/src/services/bankingService.ts
import axios from 'axios';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { API_CONFIG, createApiUrl, logApiCall, logApiResponse, logApiError } from '../config';
import store from '../store';
import { logout } from '../store/authSlice';
import { CommonActions } from '@react-navigation/native';

const API_URL = API_CONFIG.BASE_URL + '/api';

// Đọc API key từ .env file
const UNIRATE_API_KEY = 'aR7WIfkcWrio3y3n9j41PWLZ7lEy0koteWgI7L0Za5qkGjaO99EpHi53Q70ZxsjA';

// Create axios instance with auth interceptor
const apiClient = axios.create({
  baseURL: API_URL,
  timeout: API_CONFIG.TIMEOUT,
});

// Add auth token to requests
apiClient.interceptors.request.use(
  async (config) => {
    const token = await AsyncStorage.getItem('accessToken');
    if (token) {
      config.headers.Authorization = `Bearer ${token}`;
    }
    
    // Log API call
    logApiCall(config.method?.toUpperCase() || 'GET', config.url || '', config.data);
    
    return config;
  },
  (error) => {
    logApiError('REQUEST', 'unknown', error);
    return Promise.reject(error);
  }
);

// Handle auth errors
apiClient.interceptors.response.use(
  (response) => {
    // Log successful response
    logApiResponse(response.config.method?.toUpperCase() || 'GET', response.config.url || '', response.data);
    return response;
  },
  async (error) => {
    // Log error response
    logApiError(error.config?.method?.toUpperCase() || 'GET', error.config?.url || '', error);
    
    if (error.response?.status === 401) {
      // Token expired or invalid
      await AsyncStorage.removeItem('accessToken');
      store.dispatch(logout());
      // Optional: reset navigation to login
      // Nếu bạn muốn chuyển về màn hình login ngay lập tức, hãy dùng NavigationService hoặc truyền navigation vào đây
    }
    return Promise.reject(error);
  }
);

// User Profile
export async function getUserProfile() {
  const response = await apiClient.get('/user/profile');
  return response.data;
}

export async function updateUserProfile(data: any) {
  const response = await apiClient.put('/user/profile', data);
  return response.data;
}

// Account Management
export async function getAccounts() {
  const response = await apiClient.get('/banking/accounts');
  return response.data;
}

export async function getAccountDetails(accountId: string) {
  const response = await apiClient.get(`/banking/accounts/${accountId}`);
  return response.data;
}

export async function getAccountBalance(accountId: string) {
  const response = await apiClient.get(`/banking/accounts/${accountId}/balance`);
  return response.data;
}

export async function createAccount(accountData: any) {
  const response = await apiClient.post('/banking/accounts', accountData);
  return response.data;
}

// Transactions
export async function getTransactions(accountId?: string, limit = 10, offset = 0) {
  const params: any = { limit, offset };
  if (accountId) {
    params.accountId = accountId;
  }
  const response = await apiClient.get('/banking/transactions', { params });
  return response.data;
}

export async function getTransactionDetails(transactionId: string) {
  const response = await apiClient.get(`/banking/transactions/${transactionId}`);
  return response.data;
}

export async function transferMoney(transferData: {
  fromAccountId: string;
  toAccountId?: string;
  toAccountNumber?: string;
  amount: number;
  description?: string;
  transferType: 'internal' | 'external' | 'beneficiary';
}) {
  const response = await apiClient.post('/banking/transfer', transferData);
  return response.data;
}

// Verify Transfer OTP
export async function verifyTransferOtp(data: {
  transactionId: string;
  otpCode: string;
}) {
  const response = await apiClient.post('/banking/transfer/verify-otp', data);
  return response.data;
}

// Resend Transfer OTP
export async function resendTransferOtp(data: {
  transactionId: string;
}) {
  const response = await apiClient.post('/banking/transfer/resend-otp', data);
  return response.data;
}

// Cards
export async function getCards() {
  const response = await apiClient.get('/banking/cards');
  return response.data;
}

export async function getCardDetails(cardId: string) {
  const response = await apiClient.get(`/banking/cards/${cardId}`);
  return response.data;
}

// Dashboard Summary
export async function getDashboardSummary() {
  const response = await apiClient.get('/banking/dashboard/summary');
  return response.data;
}

// Account Statement
export async function getAccountStatement(accountId: string, startDate: string, endDate: string) {
  const response = await apiClient.get(`/banking/accounts/${accountId}/statement`, {
    params: { startDate, endDate }
  });
  return response.data;
}

// Beneficiaries
export async function getBeneficiaries() {
  const response = await apiClient.get('/banking/beneficiaries');
  return response.data;
}

export async function addBeneficiary(beneficiaryData: any) {
  const response = await apiClient.post('/banking/beneficiaries', beneficiaryData);
  return response.data;
}

// Kiểm tra số tài khoản đích và lấy tên chủ tài khoản
export async function checkAccountNumber(accountNumber: string) {
  const response = await apiClient.get(`/banking/accounts/lookup?accountNumber=${encodeURIComponent(accountNumber)}`);
  return response.data;
}

// Error handling helper
export function handleApiError(error: any): string {
  if (error.response?.data?.message) {
    return error.response.data.message;
  }
  if (error.message) {
    return error.message;
  }
  return 'Có lỗi xảy ra, vui lòng thử lại';
}

// User Profile APIs
export const userService = {
  // Get user profile
  getProfile: async () => {
    try {
      const response = await apiClient.get('/user/profile');
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Update user profile
  updateProfile: async (data: {
    firstName: string;
    lastName: string;
    phone?: string;
    profilePicture?: string;
  }) => {
    try {
      const response = await apiClient.put('/user/profile', data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Change password
  changePassword: async (data: {
    currentPassword: string;
    newPassword: string;
  }) => {
    try {
      const response = await apiClient.post('/user/change-password', data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Get user preferences
  getPreferences: async () => {
    try {
      const response = await apiClient.get('/user/preferences');
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Update user preferences
  updatePreferences: async (data: {
    language?: string;
    theme?: string;
    notifications?: boolean;
  }) => {
    try {
      const response = await apiClient.put('/user/preferences', data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Get KYC status
  getKycStatus: async () => {
    try {
      const response = await apiClient.get('/user/kyc-status');
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Submit KYC
  submitKyc: async (data: any) => {
    try {
      const response = await apiClient.post('/user/kyc', data);
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Upload document
  uploadDocument: async (file: any, docType: 'ID_CARD' | 'PASSPORT') => {
    try {
      const formData = new FormData();
      formData.append('file', file);
      formData.append('docType', docType);
      
      const response = await apiClient.post('/user/documents', formData, {
        headers: {
          'Content-Type': 'multipart/form-data',
        },
      });
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },

  // Deactivate account
  deactivateAccount: async () => {
    try {
      const response = await apiClient.post('/user/deactivate');
      return response.data;
    } catch (error) {
      throw handleApiError(error);
    }
  },
};

export async function updateDisplayCurrency(displayCurrency: string) {
  const response = await apiClient.put('/user/display-currency', { displayCurrency });
  return response.data;
} 

// UniRateAPI - lấy tỷ giá real-time
export async function fetchExchangeRates(base: string = 'USD') {
  try {
    console.log('🔑 [ExchangeRates] API Key available:', !!UNIRATE_API_KEY);
    console.log('🔑 [ExchangeRates] API Key length:', UNIRATE_API_KEY.length);
    
    if (!UNIRATE_API_KEY) {
      console.error('❌ [ExchangeRates] No API key found! Please set UNIRATE_API_KEY in .env file');
      throw new Error('API key not configured');
    }
    
    const url = `https://api.unirateapi.com/api/rates?api_key=${UNIRATE_API_KEY}&from=${base}`;
    console.log('🌐 [ExchangeRates] Fetching exchange rates from:', url);
    
    const response = await fetch(url);
    console.log('📡 [ExchangeRates] Response status:', response.status);
    console.log('📡 [ExchangeRates] Response ok:', response.ok);
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error('❌ [ExchangeRates] Failed to fetch exchange rates:', response.status, response.statusText);
      console.error('❌ [ExchangeRates] Error response:', errorText);
      throw new Error(`Failed to fetch exchange rates: ${response.status} ${response.statusText}`);
    }
    
    const data = await response.json();
    console.log('✅ [ExchangeRates] Raw exchange rates response:', data);
    
    // Đảm bảo trả về format đúng: { base: 'USD', rates: { VND: ..., EUR: ..., ... } }
    if (data.rates) {
      console.log('✅ [ExchangeRates] Processed exchange rates:', data.rates);
      console.log('✅ [ExchangeRates] Available currencies:', Object.keys(data.rates));
      return data;
    } else {
      console.error('❌ [ExchangeRates] Invalid exchange rates format:', data);
      throw new Error('Invalid exchange rates format');
    }
  } catch (error) {
    console.error('❌ [ExchangeRates] Error fetching exchange rates:', error);
    console.error('❌ [ExchangeRates] Error details:', error.message);
    throw error;
  }
} 

File: banking_flutter/lib/src/services/banking_service.dart
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';
import 'http_client.dart';
import '../models/account.dart';
import '../models/transaction.dart';
import '../models/beneficiary.dart';
import '../models/card.dart' as model;

class BankingService {
  final ApiClient _api;
  BankingService(this._api);

  Future<List<Account>> getAccounts({BuildContext? context}) async {
    print('🌐 Banking: Calling getAccounts...');
    final res = await _api.get(
      "${ApiConfig.banking}/accounts",
      context: context,
    );
    print('🌐 Banking: getAccounts response ${res.statusCode}: ${res.body}');
    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      // Support both flat list and wrapped shape: { data: { accounts: [] } }
      dynamic list;
      if (data is List) {
        list = data;
      } else if (data is Map &&
          data['data'] is Map &&
          (data['data']['accounts'] is List)) {
        list = data['data']['accounts'];
      } else if (data is Map && data['accounts'] is List) {
        list = data['accounts'];
      } else {
        throw Exception('Unexpected accounts response shape');
      }
      return (list as List)
          .map((e) => Account.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    throw Exception('Failed to load accounts: ${res.body}');
  }

  Future<Map<String, dynamic>> getDashboardSummary() async {
    final res = await _api.get("${ApiConfig.banking}/dashboard/summary");
    if (res.statusCode == 200) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Failed to load summary: ${res.body}');
  }

  Future<List<BankTransaction>> getTransactions({
    String? accountId,
    int limit = 10,
    int offset = 0,
    DateTime? startDate,
    DateTime? endDate,
  }) async {
    final query = <String, String>{
      'limit': '$limit',
      'offset': '$offset',
      if (accountId != null) 'accountId': accountId,
      if (startDate != null) 'startDate': startDate.toIso8601String(),
      if (endDate != null) 'endDate': endDate.toIso8601String(),
    };
    final uri = Uri.parse(
      "${ApiConfig.baseUrl}${ApiConfig.banking}/transactions",
    ).replace(queryParameters: query);

    final res = await http.get(uri, headers: await ApiClient.buildHeaders());
    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      // Support shapes:
      // 1) [ ... ]
      // 2) { data: { items: [] } }
      // 3) { data: [] }
      // 4) { items: [] }
      dynamic list;
      if (data is List) {
        list = data;
      } else if (data is Map &&
          data['data'] is Map &&
          (data['data']['items'] is List)) {
        list = data['data']['items'];
      } else if (data is Map && (data['data'] is List)) {
        list = data['data'];
      } else if (data is Map && data['items'] is List) {
        list = data['items'];
      } else if (data is Map &&
          data['data'] is Map &&
          (data['data']['transactions'] is List)) {
        list = data['data']['transactions'];
      } else {
        throw Exception('Unexpected transactions response shape');
      }
      return (list as List)
          .map((e) {
            try {
              return BankTransaction.fromJson(e as Map<String, dynamic>);
            } catch (error) {
              print('🌐 Banking: Error parsing transaction: $error');
              print('🌐 Banking: Transaction data: $e');
              return null;
            }
          })
          .where((transaction) => transaction != null)
          .cast<BankTransaction>()
          .toList();
    }
    throw Exception('Failed to load transactions: ${res.body}');
  }

  Future<List<Beneficiary>> getBeneficiaries() async {
    final res = await _api.get("${ApiConfig.banking}/beneficiaries");
    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      dynamic list;
      if (data is List) {
        list = data;
      } else if (data is Map &&
          data['data'] is Map &&
          (data['data']['beneficiaries'] is List)) {
        list = data['data']['beneficiaries'];
      } else if (data is Map && data['beneficiaries'] is List) {
        list = data['beneficiaries'];
      } else {
        throw Exception('Unexpected beneficiaries response shape');
      }
      return (list as List)
          .map((e) => Beneficiary.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    throw Exception('Failed to load beneficiaries: ${res.body}');
  }

  Future<List<model.BankCard>> getCards() async {
    final res = await _api.get("${ApiConfig.banking}/cards");
    if (res.statusCode == 200) {
      final data = jsonDecode(res.body);
      dynamic list;
      if (data is List) {
        list = data;
      } else if (data is Map &&
          data['data'] is Map &&
          (data['data']['cards'] is List)) {
        list = data['data']['cards'];
      } else if (data is Map && data['cards'] is List) {
        list = data['cards'];
      } else {
        throw Exception('Unexpected cards response shape');
      }
      return (list as List)
          .map((e) => model.BankCard.fromJson(e as Map<String, dynamic>))
          .toList();
    }
    throw Exception('Failed to load cards: ${res.body}');
  }

  Future<Map<String, dynamic>> verifyAccount(String accountNumber) async {
    print('🌐 Banking: Calling verifyAccount...');
    final res = await _api.get(
      "${ApiConfig.banking}/verify-account/$accountNumber",
    );
    print('🌐 Banking: verifyAccount response ${res.statusCode}: ${res.body}');

    if (res.statusCode == 200) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    } else if (res.statusCode == 404) {
      throw Exception('Không tìm thấy tài khoản');
    }
    throw Exception('Lỗi xác thực tài khoản: ${res.body}');
  }

  Future<Map<String, dynamic>> transfer({
    required String fromAccountId,
    required String toAccountNumber,
    required String toAccountName,
    required double amount,
    String? note,
  }) async {
    print('🌐 Banking: Calling transfer...');

    // For external transfers, we need toAccountNumber (not toAccountId)
    final body = {
      'fromAccountId': fromAccountId,
      'toAccountNumber': toAccountNumber, // Required for external transfers
      'amount': amount,
      'transferType': 'external', // External transfer by account number
      if (note != null && note.isNotEmpty) 'description': note,
    };

    final res = await _api.post("${ApiConfig.banking}/transfer", body: body);
    print('🌐 Banking: transfer response ${res.statusCode}: ${res.body}');

    if (res.statusCode == 200 || res.statusCode == 201) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Transfer failed: ${res.body}');
  }

  Future<Map<String, dynamic>> verifyTransferOtp({
    required String transactionId,
    required String otpCode,
  }) async {
    print('🌐 Banking: Calling verifyTransferOtp...');
    final body = {'transactionId': transactionId, 'otpCode': otpCode};

    final res = await _api.post(
      "${ApiConfig.banking}/transfer/verify-otp",
      body: body,
    );
    print(
      '🌐 Banking: verifyTransferOtp response ${res.statusCode}: ${res.body}',
    );

    if (res.statusCode == 200 || res.statusCode == 201) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw Exception('OTP verification failed: ${res.body}');
  }

  Future<Map<String, dynamic>> resendTransferOtp({
    required String transactionId,
  }) async {
    print('🌐 Banking: Calling resendTransferOtp...');
    final body = {'transactionId': transactionId};

    final res = await _api.post(
      "${ApiConfig.banking}/transfer/resend-otp",
      body: body,
    );
    print(
      '🌐 Banking: resendTransferOtp response ${res.statusCode}: ${res.body}',
    );

    if (res.statusCode == 200 || res.statusCode == 201) {
      return jsonDecode(res.body) as Map<String, dynamic>;
    }
    throw Exception('Resend OTP failed: ${res.body}');
  }
}
